packages/back-end/src/util/organization.util.ts

```diff
@@ -68,27 +84,167 @@ export function roleToPermissionMap(
+  newPermissions: PermissionsObject
+): PermissionsObject {
+  const updatedPermissions: PermissionsObject = { ...existingPermissions };
+
+  for (const permission in newPermissions) {
+    if (isValidPermission(permission) && newPermissions[permission] === true) {
+      updatedPermissions[permission] = true;
+    }
+  }
+
+  return updatedPermissions;
+}
+
+function mergeEnvironmentLimits(
+  existingPermissions: UserPermission,
+  newPermissions: UserPermission
+): UserPermission {
+  const existingRoleSupportsEnvLimits = hasEnvScopedPermissions(
+    existingPermissions.permissions
+  );
+  const newRoleSupportsEnvLimits = hasEnvScopedPermissions(
+    newPermissions.permissions
+  );
+
+  if (!existingRoleSupportsEnvLimits && !newRoleSupportsEnvLimits) {
+    // Neither role supports env limits, so we can skip logic below
+    return existingPermissions;
+  }
+
+  const updatedPermissions = cloneDeep(existingPermissions);
+
+  if (
+    // If the existingRole & newRole can be limited by environment
+    existingRoleSupportsEnvLimits &&
+    newRoleSupportsEnvLimits
+  ) {
+    if (
+      // and if limitAccessByEnvironment is the same for new and existing roles, we just concat the envs arrays
+      updatedPermissions.limitAccessByEnvironment ===
+      newPermissions.limitAccessByEnvironment
+    ) {
+      updatedPermissions.environments = [
+        ...new Set(
+          updatedPermissions.environments.concat(newPermissions.environments)
+        ),
+      ];
+    } else {
+      // otherwise, 1 role doesn't have limited access by environment, so it overrides the other
+      updatedPermissions.limitAccessByEnvironment = false;
+      updatedPermissions.environments = [];
+    }
+  } else {
+    // Only override existing role's env limits if the existing role doesn't support env limits, and the newRole does
+    if (!existingRoleSupportsEnvLimits && newRoleSupportsEnvLimits) {
+      updatedPermissions.limitAccessByEnvironment =
+        newPermissions.limitAccessByEnvironment;
+
+      updatedPermissions.environments = newPermissions.environments;
+    }
+  }
+  return updatedPermissions;
+}
+
+function mergeUserPermissionObj(items: UserPermission[]): UserPermission {
+  let updatedUserPermissionObj = items[0];
+
+  for (let i = 1; i < items.length; i++) {
+    updatedUserPermissionObj = mergeEnvironmentLimits(items[0], items[i]);
+    updatedUserPermissionObj.permissions = mergePermissions(
+      updatedUserPermissionObj.permissions,
+      items[i].permissions
+    );
+  }
+
+  return updatedUserPermissionObj;
+}
+
+function mergeUserAndTeamPermissions(
+  userPermissions: UserPermissions,
+  teamPermissions: UserPermissions
+) {
+  userPermissions.global = mergeUserPermissionObj([
+    userPermissions.global,
+    teamPermissions.global,
+  ]);
+
+  for (const project in teamPermissions.projects) {
+    // If the userPermissions.projects doesn't have this project, just add it
+    if (!userPermissions.projects[project]) {
```

Comment:
So if a user only has a global role of `experimenter` and is added to a team that sets them as a collaborator in project A, the user will be restricted in project A to a collaborator, right? So being added to a team may actually give you fewer permissions than you had before. Not saying this is wrong, just trying to wrap my head around it.

Reply:
@jdorn That is correct, which seems to be in alignment with what is in the app today - if a user has an `experimenter` role, but is on a team that gives them `collaborator` permissions, when we call `hasPermissions`, we check if there was a project passed in, and evaluate if they have permissions based on that project, otherwise, we fall back to the user's global role.

If we do want to change that logic, I think we would change it within `shared/permissions.utils.ts` rather than here. We would likely want to check the user's global permission, and if those were sufficient, then check the project.

Reply:
Before merging user/team permissions, fill any missing project-scoped roles with the global role.

Role:
- Global: engineer
- Project A: readonly
- Project B: undefined

Becomes:
- Global: engineer
- Project A: readonly
- Project B: engineer

Reply:
@jdorn Just pushed up [these changes](https://github.com/growthbook/growthbook/pull/1640/commits/ce8975e29f55617885f41788389b59b265a6cf66). 1 test is still failing (correctly), but I haven't been able to resolve the code issue in order to get the test to pass.

In these changes, I'm looping through all projects that the user has project-level permissions for, and if they're on a team that doesn't have permissions for that project, I'm setting the project-level permissions equal to the global role of the team.

I'm also looping through all of the projects the team has project-level permissions for, and if the user doesn't have project-level permissions for that project, I'm setting them equal to the user's global role.

However, in the case where the user doesn't have a project-level role for a project, but the team does, the user's global permissions are getting updated with that of the team's project-level permission.

Reply:
Update - I think I got the bug resolved [here](https://github.com/growthbook/growthbook/pull/1640/commits/07e5090cb3a3705966249a5697baab781f3b56d2).

Updated method:

```ts
function mergeUserAndTeamPermissions(
  userPermissions: UserPermissions,
  teamPermissions: UserPermissions
) {
  userPermissions.global = mergeUserPermissionObj([
    userPermissions.global,
    teamPermissions.global,
  ]);

  // If a user is on a team that has project-level permissions,
  // but they're user-level permissions don't have project-level permissions,
  // set the user's project-level permissions equal to their global permissions.
  for (const project in teamPermissions.projects) {
    if (!userPermissions.projects[project]) {
      userPermissions.projects[project] = {
        limitAccessByEnvironment:
          userPermissions.global.limitAccessByEnvironment,
        environments: userPermissions.global.environments,
        permissions: userPermissions.global.permissions,
      };
    }
  }

  // If the user is on a team that doesn't have project-level permissions that the user has,
  // set the team's project-level permissions equal to the team's global role
  for (const project in userPermissions.projects) {
    if (!teamPermissions.projects[project]) {
      teamPermissions.projects[project] = {
        limitAccessByEnvironment:
          teamPermissions.global.limitAccessByEnvironment,
        environments: teamPermissions.global.environments,
        permissions: teamPermissions.global.permissions,
      };
    }
  }

  // Loop through each team project and merge it's permissions with the user's project permissions
  for (const project in teamPermissions.projects) {
    userPermissions.projects[project] = mergeUserPermissionObj([
      userPermissions.projects[project],
      teamPermissions.projects[project],
    ]);
  }

  // Loop through each user project and merge it's permissions with the team's project permissions
  for (const project in userPermissions.projects) {
    userPermissions.projects[project] = mergeUserPermissionObj([
      userPermissions.projects[project],
      teamPermissions.projects[project],
    ]);
  }
}

Reply:
It feels like we could potentially merge user projects and team projects to reduce the number of loops needed.

Reply:
This doesn't look quite right.
- User has `readonly` global role
- Team has `engineer` global role and `readonly` role in Project A

I would expect the result to have `readonly` permissions in Project A.

But, I believe looking at the code above, it would result in `engineer` permissions in Project A instead.

This is because you are merging global permissions before projects.  If you move the global merge after the project loops, it should fix it.

Reply:
Ah - I see. I've updated the code and added a test case specific to this.